### ISSUE TITLE
Move away from a monorepo setup for analysis folders

### ISSUE THREAD

**Issue by Pewtro** [WoWAnalyzer/WoWAnalyzer#5049]
## Why? 

I love monorepos. I use them often, but the way it's set up in this repo makes very little sense to me. We have 57 folders in the analysis folder each being their own package, but we don't utilise the individual package for anything. We don't have a setup where we can release an individual package and have unique versioning for each package. So why is it in a monorepo package structure? It just doesn't make sense to me to introduce more complexity with that structuring to have, to my knowledge, no gain. If we are not planning to start properly utilising having multiple packages in the nearest future, then I don't see why we have separated out source code from the source code folder. I don't why we have created all those packages when all we do is build it all together at once anyways, and deploy all of it every time anyways. The entire point of monorepos is to make it easier to design and maintain microservices, but that isn't what WoWAnalyzer is - it's all deployed as one. Not to mention that having all these packages also leads to a bunch of issues with autocompletion and auto importing as seen in #4905. 

## My suggestion

Return to having class based folder structure under the src folder, so we'd have something like this (analysis folder can be called whatever): 

> src/
> ├─ analysis/
> │  ├─ class/
> │  │  ├─ spec/
> │  │  ├─ shared
> │  ├─ class2/
> │  │  ├─ spec2
> │  │  ├─ shared2

This leads to less folder bloat, it's easier to focus your work on a single area when working on a class as a whole. 

## What about classic(s)? 
As for what to do with TBC? Probably just remove it on DF branch because then WotLK is out, and then it can live together with Shadowlands on that branch. It seems incorrect to try and support multiple expansions in the same repo. But they could also just live in a tbc/ folder, the important bit is to move back into src and reduce the package/folder bloat.

**Comment by ToppleTheNun** [comment]
> As for what to do with TBC? Probably just remove it on DF branch because then WotLK is out, and then it can live together with Shadowlands on that branch. It seems incorrect to try and support multiple expansions in the same repo. But they could also just live in a tbc/ folder, the important bit is to move back into src and reduce the package/folder bloat.

If we maintain branches for each expansion (see Legion due to Legion Timewalking), would it make sense to have classic also be in its own branch?

**Comment by sMteX** [comment]
Agreed and one point I thought of - if part of the appeal of monorepo is nice import paths like `@wowanalyzer/druid`, we could achieve that with just aliases in tsconfig for example.

**Comment by Pewtro** [comment]
Personally I think it's an error that classic code was ever added to this repo, I think the games are so vastly different that it should be different repos completely, so I'm probably the wrong person to ask what to do with them, but sure it could exist in a tbc branch -- but really not the major issue I have with the structuring of the project at the moment.

**Comment by danny-janse** [comment]
Im in full support of the idea, my only question on this is if each spec/class are its own 'package' as they are now or will each class be a package? Dont think it changes alot, just my curiosity.

As for classic support, i think removing it from the dragonflight branch is a good idea. Making it its own repo makes it alot easier to maintain.  

My reasoning for this is; I have noticed that alot of things for clasic when making changes to character tab l (spirit, 3 weapon support, etc..). Thats not mentioning the expansion if-statements to disabled retail systems that arn't available on Classic,  like covenants and soulbinds. Having this makes the code more confusing.

There is also a case where i nearly renamed an ability for tbcpriest from `BINDING_HEAL` to `BINDING_HEALS` because shadowlands reworked the spell, gave it another id and name.

**Comment by Pewtro** [comment]
My proposition is that nothing will be a package, because them being packages doesn't make sense. Agree with your points on classic though :)

**Comment by danny-janse** [comment]
> My proposition is that nothing will be a package, because them being packages doesn't make sense.

That would be ideal, i've been currerntly been running into issues that certain files dont exist or changes not coming through. where i am 100% sure that i did make a change and that the files do exist. Natrually removing node_modules/.yarn* also resolves this but its a hindrance when working on the codebase non the less. I asume that them not being a package will resolve this problem?

One more argument from me to push Classic to a seperate repo instead of a branch.
I've no idea what Blizzards plans are regarding classic, where will they stop? But with the introduction of old-content Mythic+ dungeons old items will also be vaiable for retail. there will be without a doubt be problems once they start introducing even older dungeons eg with calculated stat bonuses.

**Comment by MartijnHols** [comment]
> Agreed and one point I thought of - if part of the appeal of monorepo is nice import paths like `@wowanalyzer/druid`, we could achieve that with just aliases in tsconfig for example.

The appeal of a monorepo was it being a massive step towards a future where specs could be separate repositories to make maintenance easier and give spec maintainers more freedom. If we get the interface and core versioned, spec maintainers can just install a specific version and update when they want to. No longer would we need to refactor all specs on major changes. We could also limit code access to spec repos and maybe even give more freedom on things like code style (several devs have voiced concerns about prettier). Another huge benefit is much improved code separation, where there is much less entanglement with the core, ui and class folders. The different packages have made it pretty obvious when you’re doing something you’re not supposed to while relative imports would fail at revealing that. 

I don’t know if all of that is still relevant now that WoWAnalyzer is under WCL management.

**Comment by MartijnHols** [comment]
> Personally I think it's an error that classic code was ever added to this repo, I think the games are so vastly different that it should be different repos completely, so I'm probably the wrong person to ask what to do with them, but sure it could exist in a tbc branch -- but really not the major issue I have with the structuring of the project at the moment.

Running separate apps for the different expansions would not work well from a user perspective. I reckon we should move in the opposite direction: work towards a setup where we can support any number of expansions and patches so our modules need fewer updates for future expansions. The setup in place for TBC integrated the two expansions pretty well already.

**Comment by MartijnHols** [comment]
> As for classic support, i think removing it from the dragonflight branch is a good idea. Making it its own repo makes it alot easier to maintain.
>
> My reasoning for this is; I have noticed that alot of things for clasic when making changes to character tab l (spirit, 3 weapon support, etc..). Thats not mentioning the expansion if-statements to disabled retail systems that arn't available on Classic, like covenants and soulbinds. Having this makes the code more confusing.
> 
> There is also a case where i nearly renamed an ability for tbcpriest from `BINDING_HEAL` to `BINDING_HEALS` because shadowlands reworked the spell, gave it another id and name.

I don't think this would make it easier to maintain at all. You get separate codebases for all expansions which will need to be maintained separately, will behave differently and require separate instructions to work on. I don't think anyone has time to maintain multiple branches simultaneously. So separating it would basically be cancelling it. A couple of if-statements seem like a minor inconvenience, and I think we could move a lot of the expansion specific code to expansion specific modules. From a DX POV I think this would be annoying, and, even more importantly, I reckon it would significantly hurt the UX as well as stated earlier.

As per the spells, I still believe we should kill that and switch to plain spell ids and let the core load the name/icon as needed. This is cleaner and less maintenance and allows us to better separate specs and packages. There are a few issues that we could implement some fixes for, such as WCL only having a single spells database for both TBC and retail. On the other hand, if people are already used to seeing wrong spell names/icons on WCL then it doesn't seem significant.

**Comment by Pewtro** [comment]
> > Personally I think it's an error that classic code was ever added to this repo, I think the games are so vastly different that it should be different repos completely, so I'm probably the wrong person to ask what to do with them, but sure it could exist in a tbc branch -- but really not the major issue I have with the structuring of the project at the moment.
> 
> Running separate apps for the different expansions would not work well from a user perspective. I reckon we should move in the opposite direction: work towards a setup where we can support any number of expansions and patches so our modules need fewer updates for future expansions. The setup in place for TBC integrated the two expansions pretty well already.

Can barely support retail as is, how would there ever be good support for multiple expansions?

**Comment by Pewtro** [comment]
I didn't intend for this issue to be a discussion on classic vs retail, but rather the structuring of the codebase

**Comment by Sharrq** [comment]
Yes please. I don't care what happens with classic xpacs (separate repo, separate branch, separate folder, whatever). But I very much am a huge fan of reconsolidation of the spec folders

**Comment by emallson** [comment]
> The appeal of a monorepo was it being a massive step towards a future where specs could be separate repositories to make maintenance easier and give spec maintainers more freedom. If we get the interface and core versioned, spec maintainers can just install a specific version and update when they want to. No longer would we need to refactor all specs on major changes. We could also limit code access to spec repos and maybe even give more freedom on things like code style (several devs have voiced concerns about prettier). Another huge benefit is much improved code separation, where there is much less entanglement with the core, ui and class folders. The different packages have made it pretty obvious when you’re doing something you’re not supposed to while relative imports would fail at revealing that.
> 
> I don’t know if all of that is still relevant now that WoWAnalyzer is under WCL management.

On a fundamental level, I don't see that much value in the spec packages being ultimately moved to their own repositories under independent maintainers. If those maintainers don't version bump to match potentially-breaking changes in `@wowanalyzer/core`, then they would no longer be able to be imported by the core and deployed to production (unless we allowed many cores to exist within the same production bundle, which *could* happen but would not be ideal both due to bundle size and caller/callee mismatches in shared code---I've run into the latter far too often in the Rust world to want to introduce it here). Then there are issues with abandonment (what do we do when user `A` controls `@wowanalyzer/mage-fire` but then stops responding to PRs? do we making `@wowanalyzer/mage-fire-2`? do we only use git refs for package dependencies instead of "real" npm refs?).

As far as the WCL angle: from a pragmatic perspective I think that we're interested in allowing spec contributors to have as much freedom in their maintenance as is practical. The WCL team doesn't have the bandwidth to do tons of spec support work, so our focus will always be on core features. The more that contributors can operate independently, the better.

That said: I think that there is generally a lack of interest in contributors running their own repositories. Some of that is due to the (relatively) fast turnaround for merging code here, and some of that is due to the fact that hosting your own repo comes with its own admin work that most people would rather avoid. While there is some merit (it would, for example, allow @Sharrq to have his own `prettier`-free spec repository), it also has some cost (i.e. autocompletion would remain broken in the majority of specs that weren't split into their own repositories, issues & prs are harder to track across many repositories, it becomes harder to have core people jump in and help on PRs).

Prior to making any change, I think there needs to be sufficient motivation to topple the status quo. In my mind, the issues with autocompletion are likely enough since we are already going to be making breaking changes in the DF branch.

> [insert classic v retail discussion]

The main issue IMO is the differences in core systems. It ended up not being *that* bad IMO, the major one being Classic having different stats and different talent setups. I'm happy having Classic stay in the main repo.

That said: Wrath is very likely to be the biggest WoW expansion since Legion (and possibly since OG Wrath), and I suspect that we'll get a BUNCH of people coming in and asking to help. The specs also get more complex, so there is more to do than just count your Shadowbolt casts and Curse of Elements uptime.

Definitely not going to ask existing contributors to support classic stuff, but on the WCL side I think we're interested in keeping the core working for Wrath (which should not require much work, though it could use a few extra dev utilities I think) and seeing what happens.

**Comment by Sharrq** [comment]
As much as I would love to not have to deal with prettier line length wrapping, it seems like a really really steep price to pay for it, especially for something that likely will never actually happen.

It also seems like the thing that will make everything more difficult for everyone in favor of the thing that will help like 2-3 people a little bit.

So personally I would rather can it and go back to a consolidated folder structure even if that means I have to deal with prettier wrapping line lengths for mobile phones :P

**Comment by MartijnHols** [comment]
While I don't think I can finally convince anyone of this solution, I'll still give it one last shot since it seems like such a potentially beautiful solution to me. I'll start by better describing how I envisioned it so we're actually talking about the same thing.

Glossary: I call specs "configs" since with multiple expansions we can have multiple fire mages. I imagine maybe at some time we could have multiple configs per spec, but that's a whole different discussion.

The structure I imagined would initially have all repos would be under the WoWAnalyzer organization so `@wowanalyzer/firemage` would be at https://github.com/WoWAnalyzer/firemage or https://github.com/WoWAnalyzer/magefire. That would always be step 1. Moving those to external repositories would be possible and something to consider, but not essential nor pre-decided in my mind.

If we look at what we call the "core", I think we need to split it up into multiple parts.
- @wowanalyzer/app
- @wowanalyzer/parser
- @wowanalyzer/ui
- @wownalyzer/shadowlands
- @wownalyzer/tbc
- other expansions

@wowanalyzer/app would hold all of the app code, e.g. the homepage, report selection, user interface to select player, and importantly, the code to handle loading a specific config, execute it and process results. This wouldn't be imported by any other packages, and in a weird way it doesn't import any other packages either except for the spec-configs (e.g. firemage).

@wowanalyzer/parser holds the CombatLogParser, Analyzer classes and essential modules such as Haste and SpellUsable. Those are expansion independent. The CLP works pretty much as it does now: it takes a config and some data and spits out a result. It might be worth considering moving the CLP to expansion specific packages at a later time, so your shadowlands CLP can have a selectedCombatant with covenants while the tbc CLP has its own unique stat properties, but I wouldn't do that from the start so step 1 would be a more manageable increment.

@wownalyzer/ui holds all the interface boxes that can be used by configs. Those are the simple ui elements we're used to: [BoringSpellValue](https://github.com/WoWAnalyzer/WoWAnalyzer/blob/shadowlands/src/parser/ui/BoringSpellValue.tsx), [DonutChart](https://github.com/WoWAnalyzer/WoWAnalyzer/blob/shadowlands/src/parser/ui/DonutChart.tsx) etc. We would need to consider where to locate more complicated interface items such as CastEfficiency. It could go in app so that we have just plain data coming out of the CLP and use that to show it, or it could be in the CLP as a part of the renderable results so the CLP has near full control over how the end result will appear.

@wowanalyzer/shadowlands holds shadowlands specific modules. This could be items, but we could also consider moving handling of things like covenants to modules. This seems more maintainable than building it into core and then removing it every expansion. We would never have to remove old expansions again, and new expansions could live in their own repository. Everything is separated very neatly.

We would need to version every package separately per the [semver](https://semver.org/) convention. This can be largely automated.

An important thing to consider is that the CLP has a very simple flow. It gets constructed with the config, report, combatants and profiles, then does a ton of work, and then spits out results in a single object. The input to output of this can be pre-defined and put into an interface that serves as a contract so all CLPs work the same. We have not significantly changed this flow ever, and only added a prop maybe once or twice a year.

A config would start by importing either @wownalyzer/parser or @wownalyzer/shadowlands for the CLP and the config types, analyzer etc. It would be versioned, so you can ask for v1.3.4 and you'll get that version. It might have several major feature less than v3.0.1, but it will continue to work indefinitely since the CLP's input and output remains the same.

When you get to modules, that's when you start importing @wownalyzer/ui for the interface elements. Those are the simple elements I talked about. You could use v1.3.4 which doesn't have a fancy new chart or upgrade to v2.0.1 which has a new charting library entirely and requires a small refactor of existing modules using charts.

If we do an upgrade of the core we would no longer be required to upgrade every single config at once. We could incrementally adopt new systems, new libraries as we can have some configs depend on older versions of the core until they're ready to upgrade. If you want to keep working as today, you could still require upgrades to all the existing configs before merging, but this would also open up the possibility to not do that and only do partial upgrades.

This already took an hour to consider and write up and I haven't had my coffee yet so I'm going to wrap it up here even though I feel like I said only half there is to it.

> > The appeal of a monorepo was it being a massive step towards a future where specs could be separate repositories to make maintenance easier and give spec maintainers more freedom. If we get the interface and core versioned, spec maintainers can just install a specific version and update when they want to. No longer would we need to refactor all specs on major changes. We could also limit code access to spec repos and maybe even give more freedom on things like code style (several devs have voiced concerns about prettier). Another huge benefit is much improved code separation, where there is much less entanglement with the core, ui and class folders. The different packages have made it pretty obvious when you’re doing something you’re not supposed to while relative imports would fail at revealing that.
> > I don’t know if all of that is still relevant now that WoWAnalyzer is under WCL management.
> 
> On a fundamental level, I don't see that much value in the spec packages being ultimately moved to their own repositories under independent maintainers. If those maintainers don't version bump to match potentially-breaking changes in `@wowanalyzer/core`, then they would no longer be able to be imported by the core and deployed to production (unless we allowed many cores to exist within the same production bundle, which _could_ happen but would not be ideal both due to bundle size and caller/callee mismatches in shared code---I've run into the latter far too often in the Rust world to want to introduce it here).

As per above, the way I envision it means they could be deployed to production regardless of the core version they depend on. This is basically what you meant with there being "many cores" allowed. There wouldn't be multiple ores loaded in the app, as through lazy loading only the specs that are actually being used would get loaded. For most users this means only a single "core" would be loaded. "Caller/callee mismatches" would be handled through the hard CLP interface. TypeScript does a good job preventing those issues.

> Then there are issues with abandonment (what do we do when user `A` controls `@wowanalyzer/mage-fire` but then stops responding to PRs? do we making `@wowanalyzer/mage-fire-2`? do we only use git refs for package dependencies instead of "real" npm refs?).

When the package is maintained by us we would only have one active fire mage repo. If it were to go external, which I think is an idea to take into consideration after step 1, then you would have @martijnhols/firemage and @Sharrq/firemage and we'd just switch which one we use. Whether to do git refs or npm refs, I think npm refs. There are many github actions to automatically publish and update on new versions.

> As far as the WCL angle: from a pragmatic perspective I think that we're interested in allowing spec contributors to have as much freedom in their maintenance as is practical. The WCL team doesn't have the bandwidth to do tons of spec support work, so our focus will always be on core features. The more that contributors can operate independently, the better.

I didn't mean to make this a "WCL angle" thing. I should have said: I was very actively working on making myself as obsolete as possible as I had to think about what would happen if I could no longer continue to provide support. WCL seems much more able so all this may be less essential.

> That said: I think that there is generally a lack of interest in contributors running their own repositories. Some of that is due to the (relatively) fast turnaround for merging code here, and some of that is due to the fact that hosting your own repo comes with its own admin work that most people would rather avoid. While there is some merit (it would, for example, allow @Sharrq to have his own `prettier`-free spec repository), it also has some cost (i.e. autocompletion would remain broken in the majority of specs that weren't split into their own repositories, issues & prs are harder to track across many repositories, it becomes harder to have core people jump in and help on PRs).
>
> Prior to making any change, I think there needs to be sufficient motivation to topple the status quo. In my mind, the issues with autocompletion are likely enough since we are already going to be making breaking changes in the DF branch.

What admin work? Autocompletion would work as it does for any library you import, spec maintainers have always tried to keep lists of things to do which they could do in their spec repo, PRs are indeed harder to track from a maintainer POV but giving someone maintainer rights to a config-repo would also be much easier (they only get access to a tiny part of the app), core people can still jump in and help.

> > [insert classic v retail discussion]
> 
> The main issue IMO is the differences in core systems. It ended up not being _that_ bad IMO, the major one being Classic having different stats and different talent setups. I'm happy having Classic stay in the main repo.
> 
> That said: Wrath is very likely to be the biggest WoW expansion since Legion (and possibly since OG Wrath), and I suspect that we'll get a BUNCH of people coming in and asking to help. The specs also get more complex, so there is more to do than just count your Shadowbolt casts and Curse of Elements uptime.
> 
> Definitely not going to ask existing contributors to support classic stuff, but on the WCL side I think we're interested in keeping the core working for Wrath (which should not require much work, though it could use a few extra dev utilities I think) and seeing what happens.

Nice. Just something aside from this discussion: I think it worth considering moving stuff like covenants to modules instead of combatants. If you have a `covenants` module that can use the raw combatant info and parse the covenants of that and provide a `hasCovenant` function. Continuously reworking the core is not ideal.

**Comment by emallson** [comment]
If these repositories are all going under the WoWA org, then I guess I can say: I would rather maintain 1 repository than 40+. Every added repository entails additional work. Even if the amount of work is small and infrequent, when you have 40+ it adds up.

Additionally, reaching the point that we *can* effectively split the existing monorepo along sensible lines like you've laid out will require a decent bit of work (mostly fixing imports, but still work). There is so much other stuff that we want to get done---does it really make sense for us to invest the time in that?

so I guess tl;dr I don't want to be on the hook to maintain 40 repos, and don't want to put in the work to split things into 40 repos. Reorganizing folder structure is a lot less work to put us in a place that will require less overhead long-term.
